### ISSUE TITLE
Support multi-task modules

### ISSUE THREAD

**Issue by gabe-l-hart** [caikit/caikit#246]
## Is your feature request related to a problem? Please describe.

With the introduction of [Streaming Tasks](https://github.com/caikit/caikit/pull/208), we've opened the door to having multiple `tasks` that are implemented by the same model under the hood, but different input/output data wrangling. This may also come up in the case where base text generation models can be used to accomplish other tasks (e.g. `classification`, `summarization`) using different pre/post processing.

In the current implementation, every loaded `model` supports at most one `task` which is defined in the `@module` decorator on the `module` for the `model`. This means that in a given instance of `caikit.runtime`, in order to use a given logical `model` for multiple tasks, it must be re-loaded for each task and given a different in-memory `model_id`. This is potentially very inefficient and also confusing.

## Describe the solution you'd like

The cleanest solution, in my mind, is to allow the `@module` decorator to optionally take a mapping from `task` to function name to allow the inference service binding to run functions besides `run` as the inference function for multiple different tasks.

## Describe alternatives you've considered

We could eliminate the efficiency problem problem by using a "resource" module that holds the in-memory base model and loads as a singleton and having multiple "workflow" modules that take that by reference and use it. This would still have separate `model_id` values for the different `tasks`, but would avoid the inefficiency of loading the real model multiple times.

## Additional context

* A decision on whether to implement this will have an impact on how the REST API is organized.
* If we adopt this, we'd need to expand the RPC deduction logic to handle multiple `tasks` for a given `module`

**Comment by gabe-l-hart** [comment]
Slack conversation on the topic: https://ibm-research.slack.com/archives/C0536GM5F2Q/p1694251032011869

**Comment by matthewpwilson** [comment]
+1 for this requirement. My use-case is for a multi-modal model that supports inference with multiple data types.

I wonder if `@taskmethod` could be sufficient to indicate a module supports multiple tasks? The task parameter on `@module` could then be made option.  For a multi-modal model that supports text -> image or image -> text It could look something like this:

```
@task(
    required_parameters={"text_input": str},
    output_type=ImageOutput,
)
class TextToImage(TaskBase):
    pass

@task(
    required_parameters={"image": ImageInput},
    output_type=str,
)
class ImageToText(TaskBase):
    pass

@module(
    "00110203-0405-0607-0809-0a0b02ee0e0a",
    "TextImageModule",
    "0.0.1",
)
class TextImageModule(MultiTaskModuleBase)

@TextToImage.taskmethod()
def text2image(self, input_text: str) -> ImageOutput:
    ........

@ImageToText.taskmethod()
def image2text(self, image: ImageInput) -> str:
    ........
```

`MultiTaskModuleBase` could provide a default implementation of `run()` that calls the appropriate inference method based on the task, to maintain API compatibility.
`init()`, `load()` and `save()` would be implemented in the concrete module class as normal.

**Comment by matthewpwilson** [comment]
or perhaps `@module` has a new optional `tasks: List[Type[TaskBase]]` parameter instead?

**Comment by matthewpwilson** [comment]
@gabe-l-hart @joerunde what do you think of my approach suggested above? I could potentially contribute an implementation of this or similar.

**Comment by gabe-l-hart** [comment]
@matthewpwilson Thanks for digging into this! I think there would be some fallout from this around how the `@module` decorator sets `TASK_CLASS` and how that is used in `runtime`. My gut is that we should keep `task(s)` as a property of the class so that we can perform reasoning about the supported task for a given model based on its module class. I think you could still accomplish this by making the `@TextToImage.taskmethod` decorator update the class property.

If you're game, I'd love to see a rough swing at what an implementation would look like! I think it would probably make the most sense to start with changes to `caikit.core`, but we'll definitely need to think through the implications for the `caikit.runtime` interfaces as well.

**Comment by gabe-l-hart** [comment]
Closed by #511 !
